### ISSUE TITLE
simulate combat not working for certain input data

### ISSUE THREAD

**Issue by hladkyjakub** [wesnoth/wesnoth#9556]
### Game and System Information

- **Version:** 1.18.3

### Description of the bug

simulate_combat doesn't work properly when defenders weapon is passed but no attackers weapon is passed.

### Steps to reproduce the behavior

1. into Lua console:
```
wesnoth.units.to_map({x=10,y=10,side=1,type="Elvish Archer"})
wesnoth.units.to_map({x=10,y=11,side=2,type="Master Bowman"})
a=wesnoth.units.get(10,10)
d=wesnoth.units.get(10,11)
_,_,as1,ds1 =wesnoth.simulate_combat(a,d,2) 
_,_,as2,ds2 =wesnoth.simulate_combat(a,1,d,2) 
print(as1.name,ds1.name) 
print(as2.name,ds2.name)
```
2. if works the second printed string should always be "bow", as the same defenders weapon is passed to the attack simulation (the only bad value is that first line should be both bows, becouse defender should use bow, and the attackers best ranged weapon should be selected automatically) 
the last line (sword bow) is 100% correct

### Expected behavior

My code should print:
longbow  bow
sword      bow

### Additional context

Starting point for investigation: [wesnoth-github-simulate_combat](https://github.com/wesnoth/wesnoth/blob/4f97c10e354c81ff1cac0e575bcf209fc62d3ca8/src/scripting/game_lua_kernel.cpp#L3227)

**Comment by white-haired-uncle** [comment]
> the only bad value is that first line should be both bows, becouse defender should use bow, and the attackers best ranged weapon should be selected automatically

Isn't the range in an attack chosen by the attacker?  In which case both should be "sword sword".

You seem to be saying that since the weapon for the attacker isn't specified it should be restricted by the range of the defender (when specified).  The way I read the wiki, if the attacker's weapon is not specified it will pick the best one for the attacker, and if that's a melee then there's a question of what happens to your simulated combat if you specified a ranged weapon.  I would think the answer is to use the def weap you specified if possible, and if not just ignore your input and pick the best weapon that matches the attacker's weapon range.  An alternative would be to simply have the defender be unable to retaliate.  If simulate_combat is designed to simulate combat, I'd assume the former is correct.

IMO, if simulate_combat is supposed to answer "what if ..." then you have to give it a valid scenario.

So perhaps something is broken, but it's not the first test, but the second.

**Comment by white-haired-uncle** [comment]
> My code should print:
> longbow bow

As mentioned, I think it should be "sword sword", but it should never be "longbow bow".  The attacker doesn't have a longbow, and even if it did name is the weapon type not the weapon description.

P.S.  I also looked at the average hitpoints for the various combats and according to the numbers when it says "sword bow" it really is calculating that as if it were possible (I thought maybe it would print "bow", but no damage could be inflicted since it's not a valid attack/defend combo).

**Comment by hladkyjakub** [comment]
> As mentioned, I think it should be "sword sword", but it should never be "longbow bow".  The attacker doesn't have a longbow, and even if it did name is the weapon type not the weapon description.

Attacker has a longbow. (Yes, the name is longbow)

> P.S.  I also looked at the average hitpoints for the various combats and according to the numbers when it says "sword bow" it really is calculating that as if it were possible (I thought maybe it would print "bow", but no damage could be inflicted since it's not a valid attack/defend combo).

Bow x sword is by engine 100% valid attacking, defending weapon combination.

**Comment by hladkyjakub** [comment]
> Isn't the range in an attack chosen by the attacker?  In which case both should be "sword sword".

Yes, in attack, but I want to simulate combat where attacker uses weapon of certain range.
Also the problem persists for units with 2 melee attacks (naga guard is easy to spot as better defending weapon changes based on hitpoints left)

> You seem to be saying that since the weapon for the attacker isn't specified it should be restricted by the range of the defender (when specified). 

Either yes, or another way to simulate combat, where attacker chooses best weapon with certain range.

**Comment by white-haired-uncle** [comment]
> > As mentioned, I think it should be "sword sword", but it should never be "longbow bow".  The attacker doesn't have a longbow, and even if it did name is the weapon type not the weapon description.
> 
> Attacker has a longbow. (Yes, the name is longbow)

Your attacker is an Elvish Archer, who has a weapon named bow with a description of bow.  Your defender is a Master Bowman, who has a bow with a description of longbow.  I believe the longbow will always be shown as bow based on something I read but am not going to go looking for, but I'm certain you're not going to get longbow from a unit that doesn't have one.

> 
> > P.S.  I also looked at the average hitpoints for the various combats and according to the numbers when it says "sword bow" it really is calculating that as if it were possible (I thought maybe it would print "bow", but no damage could be inflicted since it's not a valid attack/defend combo).
> 
> Bow x sword is by engine 100% valid attacking, defending weapon combination.

ranged vs melee?  I've never seen it.  Maybe it's possible, but I bet you'd have to force it.

>  I want to simulate combat where attacker uses weapon of certain range.

Run the simulation once for each type of attack and compare the results?  IDK, I think we're getting off topic here.

> Either yes, or another way to simulate combat, where attacker chooses best weapon with certain range.

Sounds like a feature request.

**Comment by hladkyjakub** [comment]
> Your attacker is an Elvish Archer, who has a weapon named bow with a description of bow.  Your defender is a Master Bowman, who has a bow with a description of longbow.  I believe the longbow will always be shown as bow based on something I read but am not going to go looking for, but I'm certain you're not going to get longbow from a unit that doesn't have one.

Fair enought, you right :D .

> ranged vs melee?  I've never seen it.  Maybe it's possible, but I bet you'd have to force it.

Yes. This one has to be forced. (I'm not working with this tho)

> >  I want to simulate combat where attacker uses weapon of certain range.
> 
> Run the simulation once for each type of attack and compare the results?  IDK, I think we're getting off topic here.

You're right, this gets off-topic.

> > Either yes, or another way to simulate combat, where attacker chooses best weapon with certain range.
> 
> Sounds like a feature request.

No, I pass simulate_combat exact weapon for defender and simulate combat completely ignores it. (Even when I say for example Naga Guard to simulate combat with shield, it's simulated for mace)

**Comment by soliton-** [comment]
Neither `simulate_combat` call seems sensible. If you don't specify an attacker weapon then the best is chosen which depends on the defender weapon thus a corresponding defender weapon will be determined as well. That should be made clearer in the documentation.

If you specify two weapons with incompatible range then whatever comes out of the simulation will presumably have little to do with anything that happens in a normal game.

**Comment by hladkyjakub** [comment]
> If you don't specify an attacker weapon then the best is chosen which depends on the defender weapon

Exactly as you say!
So make simulate_combat work to find best weapon for fixed defenders weapon.

**Comment by soliton-** [comment]
The best attacker weapon is chosen which depends on the defender weapon and thus any specified defender weapon is irrelevant while choosing the best attacker weapon.

If you want a new feature that selects an attacker weapon based on a provided defender weapon (how does that question even come up?) then we can change this into a feature request.

**Comment by gfgtdf** [comment]
If we come to the conclusion that specifying only a defender weapon is not something that we want to support anytime soon, we should make calling simulate_combat give an error if its only called with the defender weapon

**Comment by hladkyjakub** [comment]
> If we come to the conclusion that specifying only a defender weapon is not something that we want to support anytime soon, we should make calling simulate_combat give an error if its only called with the defender weapon

-_- 
...
Have you came to that conclusion?

**Comment by white-haired-uncle** [comment]
> The best attacker weapon is chosen **which depends on the defender weapon** 

I read this a lot of times, and it always struck me as referring to the _specified_ defender weapon.

But now I think I get it.  The attacker weapon is chosen, taking into account the valid (same range type, etc) weapons the defender has to choose from.  So if the specified weapon isn't a valid response, it is ignored - which is exactly what we see.  

Now I'm curious, what if the defender weapon was valid, but not the "best" according to the engine?  IOW, was the defender weapon ignored because it was not valid, or will it always be ignored when the attacker weapon is not specified?  Guess I should try it...

**Comment by gfgtdf** [comment]
> Have you came to that conclusion?

No, it was more like an answer to solitons post, it sounded a bit like implenenting this could be harder than it looked at irst.

**Comment by hladkyjakub** [comment]
> No, it was more like an answer to solitons post, it sounded a bit like implenenting this could be harder than it looked at first.
> 
> (but even _if_ we want to support it, this mostlikely won't go into wesnoth 1.18, so my comment still applies to 1.18)

Neat...
As I see it, al it would need is modifyied function of selecting best weapon, that simply doesn't take into account iteraring over defenders weapons (and instead takes just the one passed to the function) and whwn looking for best attackers weapon, it evaluates as valid only weapons of certain range (for example "ranged")

**Comment by hladkyjakub** [comment]
So no really new function would be required.

**Comment by soliton-** [comment]
> > The best attacker weapon is chosen which depends on the defender weapon and thus any specified defender weapon is irrelevant while choosing the best attacker weapon.
> > If you want a new feature that selects an attacker weapon based on a provided defender weapon (how does that question even come up?) then we can change this into a feature request.
> 
> What can I do about it? -_-

You could start with acknowledging that that is what you want and ideally motivate why you want that.

**Comment by hladkyjakub** [comment]
> You could start with acknowledging that that is what you want and ideally motivate why you want that.

My use case would be to restrict defender to one weapon, when defending against more attackers at once. So no mid-combat weapon switch is done. While making all attacks to happen on same range ("melee" or "range").
